### PR TITLE
Feature plotting

### PR DESCRIPTION
**Pull request by anaston**
This new feature extends the plotting functionalities of the toolbox.

The two main changes are:

- adding a plot_time_series function which is a very flexible way of plotting any time series data including inputs, reservoir states and labels/outputs,
- returning the model object after fitting the model on the reservoir states which allows to do some diagnostics on the model (e.g. plotting decision function or features with high model weights).

### CODE REVIEW

File: conn2res/iodata.py [hunk 1]
```diff
@@ -132,10 +139,9 @@ def fetch_dataset(task, n_trials=100, *args, **kwargs):
     if task in NEUROGYM_TASKS:
 
         # create a Dataset object from NeuroGym
-        kwargs = {'dt': 100}
```
Comment (anaston) [inline conn2res/iodata.py:135]: I removed this line and dt is passed as kwargs when calling fetch_dataset. This way other environmental variables of the ngym.Dataset can be also set (e.g. timing, cohs etc.).

Of note, dt=100 by default for several tasks including PerceptualDecisionMaking, PerceptualDecisionMakingDelayResponse, ContextDecisionMaking, so we might not even want to specify this in the examples anyway.

File: conn2res/plotting.py [hunk 2]
```diff
@@ -9,13 +9,78 @@
...
 import numpy as np
+from numpy.linalg import svd
 
 PROJ_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
 FIG_DIR = os.path.join(PROJ_DIR, 'figs')
 if not os.path.isdir(FIG_DIR):
     os.makedirs(FIG_DIR)
 
 
+class PCA:
```
Comment (anaston) [inline conn2res/plotting.py:20]: This is just a test

File: conn2res/plotting.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+import numpy as np
+from numpy.linalg import svd
+
+PROJ_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+FIG_DIR = os.path.join(PROJ_DIR, 'figs')
+if not os.path.isdir(FIG_DIR):
+    os.makedirs(FIG_DIR)
+
+
+class PCA:
```
Comment (anaston) [inline conn2res/plotting.py:20]: I created this PCA class to have a simple PCA object where we can specify which principal component we want to keep (see idx_pcs) and we can also decide to normalize the PCs by their singular values if we want so.

Perhaps it would be better to inherit this class from sklearn.decomposition.PCA but I thought it was quicker to implement my PCA from scratch than figuring out how sklearn.decomposition.PCA works. But I might have been wrong and looking backwards probably it would have been more principled to invest time understanding sklearn.decomposition.PCA properly and only extend its functionalities with a new class where needed.

File: conn2res/plotting.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        pc = (data @ self.vh.T[:, self.idx_pcs]).reshape(-1, self.n_pcs)
+
+        # normalize principal components by singular values (loop for efficiency)
+        if normalize == True:
+            for i in range(self.n_pcs):
+                pc[:, i] /= self.s[self.idx_pcs[i]]
+
+        return pc
+
+    def fit_transform(self, data, **kwargs):
```
Comment (anaston) [inline conn2res/plotting.py:76]: I tried to keep the general python nomenclature with fit/transform/fit_transform methods.

File: conn2res/plotting.py [hunk 5]
```diff
@@ -70,3 +135,77 @@ def plot_performance_curve(df, title, num=2, figsize=(12, 10), savefig=False, bl
...
                     transparent=True, bbox_inches='tight', dpi=300)
     plt.show(block=block)
+
+
+def plot_time_series(x, feature_set='orig', idx_ts=None, idx_features=None, n_features=None,
+                     scaler=1, num=1, figsize=(12, 6), subplot=None, title=None, fname='time_course',
+                     legend_label='', savefig=False, block=True, **kwargs):
+    # transform data
+    x = transform_data(
+        x, feature_set, idx_features=idx_features, n_features=n_features, scaler=scaler, **kwargs)
```
Comment (anaston) [inline conn2res/plotting.py:148]: The idea is that we transform the data before plotting so that, for instance, we can define the features which we want to plot, get principal components of the data etc.

File: conn2res/plotting.py [hunk 6]
```diff
@@ -70,3 +135,77 @@ def plot_performance_curve(df, title, num=2, figsize=(12, 10), savefig=False, bl
         fig.savefig(fname=os.path.join(FIG_DIR, f'{title}_score.png'),
                     transparent=True, bbox_inches='tight', dpi=300)
     plt.show(block=block)
+
+
+def plot_time_series(x, feature_set='orig', idx_ts=None, idx_features=None, n_features=None,
```
Comment (anaston) [inline conn2res/plotting.py]: The general idea behind this function is that we can call it multiple times when we want to create a figure. This helps a lot when we want to explore the data and see, for instance, how the inputs relate to the reservoir states etc.

If we want to plot on the same figure and/or subplot, we need to make sure to specify the same arguments (see num, subplot). We can also adjust the figure size (see figsize) depending on the number of suplots we want to have.

I tricked the way the legend works to be able to plot on the same subplot multiple times.

File: conn2res/plotting.py [hunk 7]
```diff
@@ -70,3 +135,77 @@ def plot_performance_curve(df, title, num=2, figsize=(12, 10), savefig=False, bl
...
+        idx_ts = np.arange(100)
+
+    # plot data
+    plt.plot(x[idx_ts])
+
+    # plot legend
+    legend = [f'{legend_label} {n+1}' for n in range(x.shape[1])]
+    try:  # quick fix to get previously plotted legends
+        lg = plt.gca().lines[-1].axes.get_legend()
+        legend = [text.get_text() for text in lg.texts] + legend
```
Comment (anaston) [inline conn2res/plotting.py]: This is the trick to reuse previously plotted legends.

File: conn2res/plotting.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            legend = [f'{legend_label} 1']
+        try:  # quick fix to get previously plotted legends
+            lg = plt.gca().lines[-1].axes.get_legend()
+            legend = [text.get_text() for text in lg.texts] + legend
+        except:
+            pass
+        if len(legend) <= 5:
+            plt.legend(legend, loc='upper right', fontsize=12)
+        else:
+            plt.legend(legend, loc='upper right', fontsize=12, ncol=2)
```
Comment (anaston) [inline conn2res/plotting.py:178]: For convenience, I decided to put the legend in 2 columns if there are >5 entries but obviously this can be changed.

File: conn2res/plotting.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    plt.subplot(*subplot)
+
+    # plot data
+    if x.size > 0:
+        plt.plot(x)
+
+        # add x and y limits
+        plt.xlim(xlim)
+        if ylim is not None:
+            plt.xlim(ylim)
```
Comment (anaston) [inline conn2res/plotting.py:163]: I realized xlim is a better way of plotting a subset of the signal than using using plt.plot() with an index in the signal (i.e. x[idx_ts])). This way you can navigate within the plot manually to see the rest of the signal in case you want so.

File: conn2res/plotting.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        # plt.legend(legend, loc='upper center', bbox_to_anchor=(
+        #     0.5, 1.05), fontsize=10, ncol=len(legend))
+
+    # set xtick/ythick fontsize
+    plt.xticks(fontsize=22)
+    plt.yticks(fontsize=22)
+
+    # add title
+    if title is not None:
+        plt.title(f'{title} time course', fontsize=22)
```
Comment (anaston) [inline conn2res/plotting.py:188]: The fontsizes are hard coded at the moment but probably a better solution would be to define these fontsizes (and other rc.Params outside of the function in plotting.py) so that all plotting functions can use the same settings but also allowing to change these settings easily in 1 place.

One solution could be to define a dictionary and update rc.Params but I am not really familiar with all this and I didn't have time to look this up properly. I decided to hard-code simply as when I started using rc.Params I ran into issues with setting xticks and yticks (apparently it is a matplotlib version issue but haven't really looked into it properly).

File: conn2res/plotting.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+    elif feature_set == 'decfun':
+        # calculate decision function using model fitted on time series
+        data = model.decision_function(data)
+
+    elif feature_set == 'coeff':
+        # get coefficient from model
+        if model.coef_.ndim > 1:
+            idx_class = kwargs.get('idx_class', 0)
+            coef = model.coef_[idx_class, :]
```
Comment (anaston) [inline conn2res/plotting.py:227]: In a multi-class/multi-output regression model.coef_ is 2D and I thought it is useful to plot the coefficients for all classes/outputs separately and now I set the default to plot the first class/output. But perhaps a better way would be some sort of averaging over classes/outputs.
